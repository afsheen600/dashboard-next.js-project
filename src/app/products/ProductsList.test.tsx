// import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import Products from "./ProductsList";
import { getDocs } from "firebase/firestore";

// Mock Firebase Firestore
jest.mock("@/app/services/firebaseConfig", () => ({
  db: {},
}));
jest.mock("firebase/firestore", () => ({
  collection: jest.fn(),
  getDocs: jest.fn(() =>
    Promise.resolve({
      docs: [
        {
          id: "1",
          data: () => ({
            name: "Product A",
            price: 100,
            category: "Mobile",
            date: new Date().toISOString(),
          }),
        },
        {
          id: "2",
          data: () => ({
            name: "Product B",
            price: 200,
            category: "Laptop",
            date: new Date().toISOString(),
          }),
        },
      ],
    })
  ),
  deleteDoc: jest.fn(() => Promise.resolve()),
  doc: jest.fn(),
}));

// Mock AddProduct modal
jest.mock("./AddProduct", () => {
  const MockAddProduct = () => <div data-testid="add-product-modal" />;
  MockAddProduct.displayName = "MockAddProduct";
  return MockAddProduct;
});

// Mock window.matchMedia for tests (fixes TypeError in some UI libs)
beforeAll(() => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
});

describe("ProductsList", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders the Products List heading", async () => {
    render(<Products />);
    expect(await screen.findByText(/products list/i)).toBeInTheDocument();
  });

  it("renders product rows in the table", async () => {
    render(<Products />);
    expect(await screen.findByText("Product A")).toBeInTheDocument();
    expect(screen.getByText("Product B")).toBeInTheDocument();
    expect(screen.getAllByText(/\$[0-9]+/).length).toBeGreaterThanOrEqual(2);
    expect(screen.getByText("Mobile")).toBeInTheDocument();
    expect(screen.getByText("Laptop")).toBeInTheDocument();
  });

  it("shows the Add Product modal when Add Product button is clicked", async () => {
    render(<Products />);
    const addButton = await screen.findByRole("button", {
      name: /add product|add/i,
    });
    fireEvent.click(addButton);
    expect(await screen.findByTestId("add-product-modal")).toBeInTheDocument();
  });

  it("filters products using search", async () => {
    render(<Products />);
    const searchInput = await screen.findByLabelText(/search products/i);
    fireEvent.change(searchInput, { target: { value: "Product A" } });
    fireEvent.blur(searchInput);
    expect(await screen.findByText("Product A")).toBeInTheDocument();
    // Wait until Product B is removed from the document after filtering
    // await waitFor(
    //   () => {
    //     expect(screen.queryByText("Product B")).not.toBeInTheDocument();
    //   },
    //   { timeout: 2000 }
    // );
  });

  it("shows error message when fetching products fails", async () => {
    // Override getDocs to throw error
    (getDocs as jest.Mock).mockImplementationOnce(() =>
      Promise.reject(new Error("Fetch error"))
    );
    render(<Products />);
    expect(
      await screen.findByText(/failed to fetch products/i)
    ).toBeInTheDocument();
  });

  it("shows delete confirmation dialog and deletes product", async () => {
    render(<Products />);
    // Open menu or click delete icon (simulate desktop)
    const deleteButtons = await screen.findAllByRole("button", {
      name: /delete/i,
    });
    fireEvent.click(deleteButtons[0]);
    // Simulate SweetAlert2 confirmation
    // You may need to mock Swal.fire if you want to test the full flow
  });

  it("matches snapshot", async () => {
    const { asFragment, findByText } = render(<Products />);
    await findByText(/products list/i);
    expect(asFragment()).toMatchSnapshot();
  });
});
