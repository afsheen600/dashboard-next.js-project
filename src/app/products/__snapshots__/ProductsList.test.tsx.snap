// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductsList matches snapshot 1`] = `
<DocumentFragment>
  <div />
  <div
    class="MuiBox-root css-jvose0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-2hw8qh-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBox-root css-15ubijc"
      >
        <div
          class="MuiTypography-root MuiTypography-h5 css-1w89sel-MuiTypography-root"
        >
          Products List
        </div>
        <div
          class="MuiBox-root css-h8czt7"
        >
          <div
            class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-28116y-MuiAutocomplete-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="search-products"
                id="search-products-label"
              >
                Search Products
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1rzy40g-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                  />
                </svg>
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-12tjcea-MuiInputBase-input-MuiOutlinedInput-input"
                  id="search-products"
                  role="combobox"
                  spellcheck="false"
                  type="text"
                  value=""
                />
                <div
                  class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Open"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Open"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                      data-testid="ArrowDropDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-w1u3ce"
                  >
                    <span>
                      Search Products
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-11zl0jd-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="AddCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m5 11h-4v4h-2v-4H7v-2h4V7h2v4h4z"
                />
              </svg>
            </span>
            Add Product
          </button>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-1ta4nlo-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-o1puh8"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
          role="progressbar"
          style="width: 60px; height: 60px;"
        >
          <svg
            class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20"
              stroke-width="4"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
